### PR TITLE
i#1834 memval Part 4: Adds memval sample

### LINKED ISSUE #1834: better support for clients de-referencing far or PC-relative app memory values: add drutil_insert_get_mem_value()?
Split from #1823

This seems like a more general problem than just an issue with
dr_insert_mbr_instrumentation(). While often inserted code that
de-references app memory is app code, this is not always the case, and we
want to support tool code doing so. Similarly to the stolen register on
ARM, we should at least document that the burden falls on clients to detect
and handle far refs on Linux rather than blindly copying opnds. We should
also add convenience routines if it is too awkward to work around this
using regular API routines.

A similar problem exists with PC-relative memory references: if a tool
blindly copies the operand, is DR supposed to try and mangle it,
potentially spilling a register and maybe messing up the tool's
instrumentation assumptions?

One proposal would be a drutil_insert_get_mem_value() routine in an
extension that handles the corner cases.

### PR DESCRIPTION
**Pull request by toshipiazza**
Introduces a sample client which inserts instrumentation after the
current instruction to dereference app memory, and to fill a per-thread
buffer.

Fixes #1834

### CODE REVIEW

**Comment by toshipiazza** [comment]
Note I'm adding ARM support currently, so I'm still working out some kinks. x86 works fine however

**Comment by toshipiazza** [comment]
I'm currently looking at the ARM failure in memval_simple, but it seems to be a problem with how I'm using `dr_insert_clean_call`. If I apply the following patch,

```diff
diff --git a/ext/drx/drx_buf.c b/ext/drx/drx_buf.c
index 0caf226..ca28fd3 100644
--- a/ext/drx/drx_buf.c
+++ b/ext/drx/drx_buf.c
@@ -696,6 +696,7 @@ safe_memcpy(drx_buf_t *buf, void *src, size_t len)
     size_t written;
     bool ok;

+    dr_printf("safe_memcpy(buf: %p, src: %p, len: %d)\n", buf, src, len);
     DR_ASSERT_MSG(buf->buf_size >= len,
                   "buffer was too small to fit requested memcpy() operation");
     /* try to perform a safe memcpy */
@@ -711,6 +712,12 @@ safe_memcpy(drx_buf_t *buf, void *src, size_t len)
     BUF_PTR(data->seg_base, buf->tls_offs) += len;
 }

+static void
+dump_arg(void *foo)
+{
+    dr_printf("%p\n", foo);
+}
+
 DR_EXPORT
 void
 drx_buf_insert_buf_memcpy(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
@@ -722,6 +729,9 @@ drx_buf_insert_buf_memcpy(void *drcontext, drx_buf_t *buf, instrlist_t *ilist,
         opnd_t buf_opnd = OPND_CREATE_INTPTR(buf);
         opnd_t src_opnd = opnd_create_reg(src);
         opnd_t len_opnd = OPND_CREATE_INTPTR((short)len);
+        dr_insert_clean_call(drcontext, ilist, where, (void *)dump_arg, false, 1, buf_opnd);
+        dr_insert_clean_call(drcontext, ilist, where, (void *)dump_arg, false, 1, src_opnd);
+        dr_insert_clean_call(drcontext, ilist, where, (void *)dump_arg, false, 1, len_opnd);
         dr_insert_clean_call(drcontext, ilist, where, (void *)safe_memcpy, false, 3,
                              buf_opnd, src_opnd, len_opnd);
     } else {
```

I see the following behaviour:

```
./bin32/drrun -c ./api/bin/libmemval_simple.so -- ls
...
0x4c4d9120 # buf
0x7ecbff54 # src
0x00000014 # len
safe_memcpy(buf: 0x4c4d9120, src: 0x00000014, len: 20)
<(1+x) Handling our fault in a TRY at 0x54da7694>
<Application /bin/ls (14389).  DynamoRIO Sample Client 'memval_simple' internal crash at PC 0x54da7694.  Please report this at http://dynamorio.org/issues.  Program aborted.
Received SIGSEGV at pc 0x54da7694 in thread 14389
...
```

We're clearly crashing because `src` is 0x14 (the same as the size), but if we look at the output from the clean calls earlier we see that `src` is clearly a valid pointer! Even more mysteriously, I don't see this problem in the drx_buf test (it passes for me). What's going on?

**Comment by derekbruening** [comment]
Is there a bug in ARM's clean call impl for 3 args?  Maybe all the tests of it only pass 1 or 2 args?

**Comment by toshipiazza** [comment]
Hmm, when looking through the AArchXX implementation of `insert_parameter_preparation()`, I see the following comment:

```c
/* FIXME i#1551, i#1569: we only implement naive parameter preparation,
 * where args are all regs or immeds and do not conflict with param regs.
 */
```

If I remember correctly, memval_simple crashed on the second usage of `drx_buf_insert_buf_memcpy()`, which used `r2` as the src register. This also corresponds to the third argument register (i.e. len?). Will double check that this was the case.

File: api/samples/memval_simple.c [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+
+/* Requires that hex_buf be at least as long as 2*memref->size + 1. */
+static char *
+write_hexdump(char *hex_buf, byte *write_base, mem_ref_t *mem_ref)
+{
+    int i;
+    char *hexstring = hex_buf, *needle = hex_buf;
+
+    for (i = mem_ref->size - 1; i >= 0; --i)
+        needle += dr_snprintf(needle, 2*mem_ref->size+1-2*i, "%02x", write_base[i]);
```
Comment (derekbruening) [inline api/samples/memval_simple.c]: Safer to use the returned value for the size: so 2*mem_ref->size+1-(needle-hex_buf)

File: api/samples/memval_simple.c [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+    /* find the largest necessary buffer so we only perform a single allocation */
+    for (mem_ref = trace_base; mem_ref < trace_ptr; mem_ref++) {
+        if (mem_ref->size > largest_size)
+            largest_size = mem_ref->size;
+    }
+    hex_buf = dr_thread_alloc(drcontext, 2*largest_size+1);
+    /* write the memrefs to disk */
+    for (mem_ref = trace_base; mem_ref < trace_ptr; mem_ref++) {
+        /* Each memref in the trace buffer has an "associated" write in the write buffer.
+         * We pull mem_reg->size bytes from the write buffer, and assert we haven't yet
```
Comment (derekbruening) [inline api/samples/memval_simple.c]: s/mem_reg/mem_ref/

File: api/samples/memval_simple.c [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+static reg_id_t
+instrument_mem(void *drcontext, instrlist_t *ilist, instr_t *where, opnd_t ref)
+{
+    reg_id_t reg_ptr, reg_tmp, reg_addr;
+    ushort type, size;
+    bool ok;
+
+    if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_tmp)
+            != DRREG_SUCCESS) {
```
Comment (derekbruening) [inline api/samples/memval_simple.c]: style: indent should be under 'd' of 'drreg' -- but prob fits on line above?  I wish github let me set a max line width and show a marker there.

File: api/samples/memval_simple.c [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+        DR_ASSERT(false);
+        return DR_REG_NULL;
+    }
+    if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_ptr)
+            != DRREG_SUCCESS) {
```
Comment (derekbruening) [inline api/samples/memval_simple.c]: Ditto

File: api/samples/memval_simple.c [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+    /* inserts memref addr */
+    drx_buf_insert_buf_store(drcontext, trace_buffer, ilist, where, reg_ptr, DR_REG_NULL,
+                             opnd_create_reg(reg_tmp), OPSZ_PTR,
+                             offsetof(mem_ref_t, addr));
+    if (IF_AARCHXX_ELSE(true, false)) {
+        /* At this point we save the write address for later, because reg_tmp's value
+         * will get clobbered on ARM.
+         */
+        if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_addr)
+                != DRREG_SUCCESS) {
```
Comment (derekbruening) [inline api/samples/memval_simple.c]: Ditto

File: api/samples/memval_simple.c [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+
+static void
+instrument_post_write(void *drcontext, instrlist_t *ilist, instr_t *where, opnd_t memref,
+                      instr_t *write, reg_id_t reg_addr)
+{
+    reg_id_t reg_ptr;
+    ushort stride = (ushort)drutil_opnd_mem_size_in_bytes(memref, write);
+
+    if (drreg_reserve_register(drcontext, ilist, where, NULL, &reg_ptr)
+            != DRREG_SUCCESS) {
```
Comment (derekbruening) [inline api/samples/memval_simple.c]: Ditto

File: api/samples/memval_simple.c [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+ * DAMAGE.
+ */
+
+/* Code Manipulation API Sample:
+ * memval_simple.c
+ *
+ * Records and dumps app write addresses, and their corresponding written values.
+ *
+ * (1) It fills two per-thread-buffers with inlined instrumentation.
+ * (2) Once the buffer has been filled up, a fault handler will redirect execution
```
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

File: api/samples/memval_simple.c [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+ *     to our trace buffer handler, where we dump the memrefs to disk.
+ *
+ * This sample illustrates
+ * - inserting instrumentation after the current instruction to read the value
+ *   written by it.
+ * - the use of drutil_expand_rep_string() to expand string loops to obtain
+ *   every memory reference,
```
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

File: api/samples/memval_simple.c (hunk 1 above)
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

File: api/samples/memval_simple.c (hunk 2 above)
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

File: api/samples/memval_simple.c (hunk 3 above)
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done, I checked and it doesn't fit (by one char).

File: api/samples/memval_simple.c (hunk 4 above)
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

File: api/samples/memval_simple.c (hunk 5 above)
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

File: api/samples/memval_simple.c (hunk 6 above)
Comment (toshipiazza) [inline api/samples/memval_simple.c]: Done.

**Comment by toshipiazza** [comment]
Should I wait for the FIXME in `dr_insert_clean_call` to be resolved (and should I be the one to solve it?) before I merge in my changes? If I merge these in, then we'll see drx_buf-test failing on ARM.

**Comment by derekbruening** [comment]
> Should I wait for the FIXME in dr_insert_clean_call to be resolved (and should I be the one to solve it?) before I merge in my changes? If I merge these in, then we'll see drx_buf-test failing on ARM.

It would be great if you fixed it :)   (Via separate pull request.)  Or you could disable for ARM the part of drx_buf-test that will fail and in drx.h mark the routine that uses the clean call as NYI on ARM.

**Comment by derekbruening** [comment]
For future pull requests, please take a look at the new workflow: we're using feature branches in the DR repo itself now, so no private clone.

**Comment by toshipiazza** [comment]
With regards to maybe fixing up `dr_insert_clean_call`, I'll do more analysis and see what needs to be done (probably before the end of the week) before determining whether I want to just disable ARM here or address the FIXME in a separate PR.

**Comment by toshipiazza** [comment]
FYI Because I'm a little busy with school atm, I'll just disable the build of memval_simple on ARM for now, and later this week I'll look into cleaning up `dr_insert_clean_call` to some degree.

**Comment by toshipiazza** [comment]
Hmm it's been a couple hours and travis says all tests passed [here](https://travis-ci.org/DynamoRIO/dynamorio/builds/204075231) but it looks like the github UI says it hasn't yet finished. Is there a problem or is there just some lag?

**Comment by derekbruening** [comment]
I have never seen this before.  I've seen the Mac build sit in a queue waiting for a free Travis machine for hours, but here Mac has run.  I'll go ahead and override.  Maybe we should file a bug against Travis.

**Pull request merged**